internal/transport/http2_server.go

```diff
@@ -371,32 +371,43 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 			timeoutSet = true
 			var err error
 			if timeout, err = decodeTimeout(hf.Value); err != nil {
-				headerError = true
+				headerError = err
 			}
 		default:
 			if isReservedHeader(hf.Name) && !isWhitelistedHeader(hf.Name) {
 				break
 			}
 			v, err := decodeMetadataHeader(hf.Name, hf.Value)
 			if err != nil {
-				headerError = true
+				headerError = err
 				logger.Warningf("Failed to decode metadata header (%q, %q): %v", hf.Name, hf.Value, err)
 				break
 			}
 			mdata[hf.Name] = append(mdata[hf.Name], v)
 		}
 	}
 
-	if !isGRPC || headerError {
-		t.controlBuf.put(&cleanupStream{
+	if !isGRPC {
+		t.controlBuf.put(&headerFrame{
 			streamID: streamID,
-			rst:      true,
-			rstCode:  http2.ErrCodeProtocol,
-			onWrite:  func() {},
+			hf: []hpack.HeaderField{
+				{Name: ":status", Value: "415"},
+			},
+			endStream: true,
 		})
 		return false
 	}
 
+	if headerError != nil {
+		if err := t.WriteStatus(
```

Comment:
Similar Q regarding whether RST_STREAM happens here automatically.

Reply:
`WriteStatus` can trigger an RST but only if the stream's status is `streamActive`. Should I check if the status is stream active, and send an RST_STREAM if the status is not `streamActive`?

Reply:
This is the initial header for the stream, right?  So it won't be active unless this initializes it that way?  We should make sure the RST_STREAM is always sent, since we won't be processing any further frames for this stream ID.